M2/Macaulay2/d/pthread.d

```diff
@@ -231,6 +232,20 @@ export getIOThreadMode(e:Expr):Expr := (
     else WrongArg("a file or ()"));
 setupfun("getIOThreadMode", getIOThreadMode);
 
+-- TODO: what if we cancel a task before it unlocks the mutex?
+mutex := newMutex;
+threadLock(c:Code):Expr := (
+    r := 0;
+    while (r = trylock(mutex); r == Ccode(int, "EBUSY")) do (
```

Comment:
This is not good. I think you want `pthread_cond_wait` or similar.

Reply:
From what I've read, `pthread_mutex_trylock` is the only way to check if a mutex is locked.  `pthread_cond_wait` blocks until some condition is met, and we don't want to block since we want to check for interrupts.

Reply:
I definitely think we need pthread conditions, because you can signal only one or all threads using pthread_cond_signal and pthread_cond_broadcast that an event has changed. This is how tasks are implemented too. e.g. this is how `taskResult` blocks:
https://github.com/Macaulay2/M2/blob/2565455411d15a3386204aa62a00e20ee5c0e99f/M2/Macaulay2/system/supervisor.cpp#L210-L215
In fact, I suspect if we change `interrupt_handler` in `main.cpp` to also signal the current blocking task, it would interrupt just fine.